src/Compilers/Server/VBCSCompilerTests/ClientConnectionHandlerTests.cs

```diff
@@ -0,0 +1,182 @@
+{
+    public class ClientConnectionHandlerTests
+    {
+        [Fact]
+        public async Task ThrowDuringBuild()
+        {
+            var compilerServerHost = new TestableCompilerServerHost(delegate { throw new Exception(); });
+            var clientConnectionHandler = new ClientConnectionHandler(compilerServerHost);
+            var clientConnection = new TestableClientConnection()
+            {
+                ReadBuildRequestFunc = _ => Task.FromResult(ProtocolUtil.EmptyCSharpBuildRequest),
+            };
+            var completionData = await clientConnectionHandler.ProcessAsync(Task.FromResult<IClientConnection>(clientConnection)).ConfigureAwait(false);
+            Assert.Equal(CompletionData.RequestError, completionData);
+        }
+
+        [Fact]
+        public async Task ThrowReadingRequest()
+        {
+            var compilerServerHost = new TestableCompilerServerHost(delegate
+            {
+                Assert.True(false, "Should not reach compilation");
+                throw new Exception("");
+            });
+            var clientConnectionHandler = new ClientConnectionHandler(compilerServerHost);
+            var clientConnection = new TestableClientConnection()
+            {
+                ReadBuildRequestFunc = _ => throw new Exception(),
+            };
+
+            var completionData = await clientConnectionHandler.ProcessAsync(Task.FromResult<IClientConnection>(clientConnection)).ConfigureAwait(false);
+            Assert.Equal(CompletionData.RequestError, completionData);
+        }
+
+        [Fact]
+        public async Task ThrowWritingResponse()
+        {
+            var compilerServerHost = new TestableCompilerServerHost(delegate { return ProtocolUtil.EmptyBuildResponse; });
+            var clientConnectionHandler = new ClientConnectionHandler(compilerServerHost);
+            var threwException = false;
+            var clientConnection = new TestableClientConnection()
+            {
+                ReadBuildRequestFunc = _ => Task.FromResult(ProtocolUtil.EmptyCSharpBuildRequest),
+                WriteBuildResponseFunc = (response, cancellationToken) =>
+                {
+                    threwException = true;
+                    throw new Exception("");
+                }
+            };
+
+            var completionData = await clientConnectionHandler.ProcessAsync(Task.FromResult<IClientConnection>(clientConnection)).ConfigureAwait(false);
+            Assert.Equal(CompletionData.RequestError, completionData);
+            Assert.True(threwException);
+        }
+
+        /// <summary>
+        /// Make sure that when compilation requests are disallowed we don't actually process them
+        /// </summary>
+        [Fact]
+        public async Task CompilationsDisallowed()
+        {
+            var compilerServerHost = new TestableCompilerServerHost(delegate
+            {
+                Assert.True(false, "Should not reach compilation when compilations are disallowed");
```

Comment:
I'm confused by this. An assert failure just throws an exception. Aren't you trying to assert that this method was never called? In that case you probably need to declare a flag in the outer scope, capture and set it in here, then assert that it was not set, right?

Reply:
good catch